### ISSUE TITLE
Bug report: jparse -J 3 does not correctly print JSON number values

### ISSUE THREAD

**Issue by lcn2** [ioccc-src/mkiocccentry#785]
## Describe the bug

When `-J 3` is in effect, JSON numbers are printed as "false" instead of the string containing their value

## To Reproduce

Steps to reproduce the behavior (adjust steps as needed):

0. ./jparse/jparse -J 3 jparse/test_jparse/test_JSON/good/two-e0.json
1. observe the output:

```

JSON tree[3]:	lvl: 0	type: JTYPE_NUMBER	{converted,parsed}: false
in parse_json(): JSON debug[1]: valid JSON
valid JSON
```

## Expected behavior

The JSON number as a string (i.e., `as_str`) is printed instead of "false".

## Relevant Output

Consider the "type: JTYPE_NUMBER" line output from commands as well:

```sh
./jparse/jparse -J 3 jparse/test_jparse/test_JSON/good/42.json
./jparse/jparse -J 3 jparse/test_jparse/test_JSON/good/top_level_array.json
```

## bug report attachment

We request that you **PLEASE** run the command:

```sh
./bug_report.sh
```

and attach the bug report file that `./bug_report.sh` generated.
The bug report file will be a filename in the same directory with a
filename of the form: `bug-report.YYYYMMDD.HHMMSS.txt` and it will tell you the
name as well.

vvvvvvvvvvvvvvvvvvvvvvvvv --- Drag the file into the area below, please!

[bug-report.20230720.133326.txt](https://github.com/ioccc-src/mkiocccentry/files/12113854/bug-report.20230720.133326.txt)

^^^^^^^^^^^^^^^^^^^^^^^^^ --- Drag the file into the area above, please!

## Please provide the following information:

 - Did you attach the above requested bug report file?  yes
 - What is the airspeed Velocity of an Unladen Swallow? see http://style.org/unladenswallow/

## Additional context

The reason to print the `as_str` as the "value" for a NUMBER is that one may avoid issues related to the machine hardware representation, or lack of representation of the "value".

### Suggested enhancement

The text inside "{...}" for "JTYPE_NUMBER" is needlessly verbose and simply refers to two of the several `struct json_number` booleans that pertain to a JSON NUMBER: all be it two of the less likely to be needed booleans associated with `struct json_number`.

The `-J 3` output is intended to be for debugging purposes and is intended for general use.  Such `-J` output is intended for those who are familiar with JSON parser structure internals.

We recommend a more compact set of letters be used, one per useful boolean, for `-J 3` output related to `struct json_number`.

We recommend that `vjson_fprint()` be modified to print within "{...}" compact letter symbols, comma separated, for `JTYPE_NUMBER` in a similar way to that of `JTYPE_STRING`.

- p for parsed
- c for converted
- \- for is_negative
- F for is_floating
- E for is_e_notation
- I for is_integer
- 8 for int8_sized
- u8 for uint8_sized
- 16 for int16_sized
- u16 for uint16_sized
- 32 for int32_sized
- u32 for uint32_sized
- 64 for int64_sized
- u64 for uint64_sized
- i for int_sized
- ui for uint_sized
- l for long_sized
- ul for ulong_sized
- ll for longlong_sized
- ull for ulonglong_sized
- ss for ssize_sized
- s for size_sized
- o for off_sized
- m for maxint_sized
- um for umaxint_sized
- f for float_sized
- fi for as_float_int
- d for double_sized
- di for as_double_int
- ld for longdouble_sized
- ldi for as_longdouble_int

The order of the "command separated boolean flags" should be the same order as found in the `struct json_number` structure.

The entire set of values need not be printed at JSON debug level 3.  For example, at `-J 3` if `int8_sized` is true, one may assume that `int16_sized` and `uint16_sized` and `int32_sized` and ... `uint64_sized` are all true.  So at `-J 3` when printing ",8" one may skip printing ",16" and skip ",u16" and skip ",32" and skip ... and skip ",u64".

A similar reduction for "long_sized" skipping at `-J 3`, printing for "longlong_sized" and "ulonglong_sized".  And "float_sized" for skipping at `-J 3`, printing for "double_sized" .. "longdouble_sized".

Also when `is_negative` is false and `int8_sized` is true, then `uint8_sized` is also true.  So without `is_negative` ate `-J 3`, one may print ",u8" and skip ",8".

A similar reduction for  signed vis unsigned for floating point may be made at `-J 3`.

At `-J 3` if we indicate the value is, say, an `int8_sized` integer, we could assume the value was both `converted` and `parsed`:

```
... type: JTYPE_NUMBER	{-,I,8}: -2
```

Of course, at `-J 3` when if `parsed` is  false and `converted` is false, then perhaps printing:

```
... type: JTYPE_NUMBER	{}: ((NULL))
```

is best.

When `parsed` is  false, the `as_str` structure member is probably a NULL pointer.

BTW: One could argue that nothing should be printed after the "}: " in the case of a NULL pointer.  On the other hand _((NULL))_ is also common as so we recommend that.

And at `-J 3` when is `converted` is false, while `parsed` is true, then perhaps printing:

```
... type: JTYPE_NUMBER	{p,E,-}: -1e1000000000000000000000000000
```

is a good idea.

In short, at `-J 3` print the minimal comma separated set of boolean indicators between "{...}".

We recommend that at `-J 7` and higher, all booleans be printed and that no reductions be made.

## UPDATE 0

In the **BOGUS** case where `parsed` is false but `converted` is true, and nothing else is true, then perhaps:

```
... type: JTYPE_NUMBER	{c}: ((NULL))
```

is best?

## UPDATE 1

In connection with [comment 1644761699](https://github.com/ioccc-src/mkiocccentry/issues/785#issuecomment-1644761699), we recommend that `parsed` be moved above the `converted` boolean for all of the "struct json_foo" structures in `jparse/json_parse.h`.

**Comment by xexyl** [comment]
I will look at this tomorrow morning. Please assign it to me.

**Comment by lcn2** [comment]
We encounter this problem while trying to figure out JSON parse tree levels.  Looking at the "as_str" values printed was a good way to try an figure out what levels are what.

The enhancement part of this bug report (which also made this issue an **enhancement**) was noticed when we saw that most of the `struct json_number` booleans were missing.

The  **enhancement** part of this issue suggests that at `-J 3` the printing of comma separated boolean indicators could be optimized, so for the JSON number:

```json
-2
```

one could, for `-J 3`, print:

```
... type: JTYPE_NUMBER	{-,I,8}: -2
```

At `-J 5` print everything:

```
... type: JTYPE_NUMBER	{c,p,-,I,8,16,32,64,i,l,ll,o,ss,m,f,fi,d,di,ld,ldi}: -2
```

We recommend that the order of "comma separated boolean indicators" within "{...}" follow the order in the `struct json_number` structure.  This is because more than likely, someone who is looking at such indicators is probably looking at the `struct json_number` structure in `jparse/json_parse.h` as well.  :-)

**Comment by xexyl** [comment]
.. actually looking at the first part of the report I know exactly what this is. It originally did print true/false but I modified it to show if parsed was also true or false. This should not be a problem but the other parts will take more time of course. I'll now reply to the comments ...

**Comment by xexyl** [comment]
... actually this is a lot to process right now so I'll reply to other things first. I might not reply to this stuff until I'm more awake. I do have some thoughts but I'm not awake enough to process them all.

**Comment by xexyl** [comment]
The funny thing is is that I was trying to figure out why in jval and jnamval this did not report an unconverted number. There was a bug in the function that prints this out. So...the bug for this number is fixed and I will do a commit now. I then have to go afk and probably for a fair bit of time.

This does not change the enhancement status of this of course ... anyway I'll bbl and hopefully I can deal with everything. If not maybe tomorrow. I hope today though and I should have more time today before I leave so it'd be good if I could spend time here too.

**Comment by xexyl** [comment]
Commit https://github.com/ioccc-src/mkiocccentry/pull/786/commits/81b7140af06d15674b0ecdaceed07617087ec027 fixes at least one bug with this issue: that number should be representable though I'm not sure how exactly (but I'm not looking at the limits of numbers either and I don't know in particular floating point whereas I know some of the max/min for ints).

This does not resolve the enhancements of course but once this is merged you can at least use that file as you wanted to.

Have to go afk a while .. hopefully I can do more later on but we'll see.

**Comment by xexyl** [comment]
> ## UPDATE 1
> In connection with [comment 1644761699](https://github.com/ioccc-src/mkiocccentry/issues/785#issuecomment-1644761699), we recommend that `parsed` be moved above the `converted` boolean for all of the "struct json_foo" structures in `jparse/json_parse.h`.

Done in https://github.com/ioccc-src/mkiocccentry/pull/787/commits/46ab576f17d99f10f4ae2336a29997fc8f8a81f5.

Really going afk now .. will hopefully be able to do more in a while ...

**Comment by lcn2** [comment]
As of commit e22cdf1ede890d126fd65dab1cae57f8088afec0 (Merge pull request #788):

The output of:

```sh
./jparse/jparse -J 3 jparse/test_jparse/test_JSON/good/top_level_array.json 2>&1 | grep JTYPE_NUMBER
```

currently produces:

```
JSON tree[3]:	lvl: 1	type: JTYPE_NUMBER	val+16: 21701
JSON tree[3]:	lvl: 1	type: JTYPE_NUMBER	f+val: 2.718000	floating: false
JSON tree[3]:	lvl: 1	type: JTYPE_NUMBER	f+val: 6.023000e+23	e-notation floating point: true
JSON tree[3]:	lvl: 2	type: JTYPE_NUMBER	val+8: 1
JSON tree[3]:	lvl: 2	type: JTYPE_NUMBER	val+8: 2
JSON tree[3]:	lvl: 2	type: JTYPE_NUMBER	val+8: 3
JSON tree[3]:	lvl: 2	type: JTYPE_NUMBER	f+val: 6.674300e-11	e-notation floating point: false
JSON tree[3]:	lvl: 3	type: JTYPE_NUMBER	val+8: 4
JSON tree[3]:	lvl: 3	type: JTYPE_NUMBER	val+8: 5
JSON tree[3]:	lvl: 3	type: JTYPE_NUMBER	val+8: 6
```

Perhaps this would be better at `-J 3`:

```
JSON tree[3]:	lvl: 1	type: JTYPE_NUMBER	value{I,16}: 21701
JSON tree[3]:	lvl: 1	type: JTYPE_NUMBER	value{F,f}: 2.718000	floating: false
JSON tree[3]:	lvl: 1	type: JTYPE_NUMBER	value{E,d,di}: 6.023000e+23	e-notation floating point: true
JSON tree[3]:	lvl: 2	type: JTYPE_NUMBER	value{I,8}: 1
JSON tree[3]:	lvl: 2	type: JTYPE_NUMBER	value{I,8}: 2
JSON tree[3]:	lvl: 2	type: JTYPE_NUMBER	value{I,8}: 3
JSON tree[3]:	lvl: 2	type: JTYPE_NUMBER	value{Elf}: 6.674300e-11	e-notation floating point: false
JSON tree[3]:	lvl: 3	type: JTYPE_NUMBER	value{I,8}: 4
JSON tree[3]:	lvl: 3	type: JTYPE_NUMBER	value{I,8}: 5
JSON tree[3]:	lvl: 3	type: JTYPE_NUMBER	value{I,8}: 6
```

or something like that.

**Comment by xexyl** [comment]
> As of commit [e22cdf1](https://github.com/ioccc-src/mkiocccentry/commit/e22cdf1ede890d126fd65dab1cae57f8088afec0) (Merge pull request #788):
> 
> The output of:
> 
> ```shell
> ./jparse/jparse -J 3 jparse/test_jparse/test_JSON/good/top_level_array.json 2>&1 | grep JTYPE_NUMBER
> ```
> 
> currently produces:
> 
> ```
> JSON tree[3]:	lvl: 1	type: JTYPE_NUMBER	val+16: 21701
> JSON tree[3]:	lvl: 1	type: JTYPE_NUMBER	f+val: 2.718000	floating: false
> JSON tree[3]:	lvl: 1	type: JTYPE_NUMBER	f+val: 6.023000e+23	e-notation floating point: true
> JSON tree[3]:	lvl: 2	type: JTYPE_NUMBER	val+8: 1
> JSON tree[3]:	lvl: 2	type: JTYPE_NUMBER	val+8: 2
> JSON tree[3]:	lvl: 2	type: JTYPE_NUMBER	val+8: 3
> JSON tree[3]:	lvl: 2	type: JTYPE_NUMBER	f+val: 6.674300e-11	e-notation floating point: false
> JSON tree[3]:	lvl: 3	type: JTYPE_NUMBER	val+8: 4
> JSON tree[3]:	lvl: 3	type: JTYPE_NUMBER	val+8: 5
> JSON tree[3]:	lvl: 3	type: JTYPE_NUMBER	val+8: 6
> ```
> 
> Perhaps this would be better at `-J 3`:
> 
> ```
> JSON tree[3]:	lvl: 1	type: JTYPE_NUMBER	value{I,16}: 21701
> JSON tree[3]:	lvl: 1	type: JTYPE_NUMBER	value{F,f}: 2.718000
> JSON tree[3]:	lvl: 1	type: JTYPE_NUMBER	value{E,d,di}: 6.023000e+23
> JSON tree[3]:	lvl: 2	type: JTYPE_NUMBER	value{I,8}: 1
> JSON tree[3]:	lvl: 2	type: JTYPE_NUMBER	value{I,8}: 2
> JSON tree[3]:	lvl: 2	type: JTYPE_NUMBER	value{I,8}: 3
> JSON tree[3]:	lvl: 2	type: JTYPE_NUMBER	value{E,f}: 6.674300e-11
> JSON tree[3]:	lvl: 3	type: JTYPE_NUMBER	value{I,8}: 4
> JSON tree[3]:	lvl: 3	type: JTYPE_NUMBER	value{I,8}: 5
> JSON tree[3]:	lvl: 3	type: JTYPE_NUMBER	value{I,8}: 6
> ```
> 
> or something useful like that for `-J 3`. What do you think?

This is a **_very_** helpful example: thanks! This was one of the issues I had: no example to confirm what you had in mind. 

I slept in till 3 or 3:30 today so I'm hoping that I can get more done though right now I'm not awake enough work on _this_ issue. I did fix an issue with `jnamval` based on a change you made in the `json_util_README.md` file and opened a pull request a few minutes ago.

**Comment by xexyl** [comment]
I am currently on the exercise bike and then I have a couple other things to do. After that I am hoping to look at this issue at least for a bit. Maybe I can reply to comments if not start working on it. 

I am feeling like today I won't get a lot done but I hope to get some done anyway: besides the fixes I made earlier. 

Probably will be a day that I just take it easy but hopefully that won't be the entire weekend. I am sure that I will do more tomorrow if not today. I did manage to sleep in today and I went to sleep earlier too but it was a difficult night also. 

Either way I expect that if I don't get any more code in that I will at least be able to reply to comments. 

I would do it with the phone but that gets to be really difficult and annoying so I don't think I will try. 

Just wanted to give you an update. 

Hope you had a great night!

**Comment by xexyl** [comment]
QUESTION: should the test script check for exact output of debug level messages too?

I am not sure. I think maybe not: depending on the system it would possibly be different output.

**Comment by xexyl** [comment]
> QUESTION: should the test script check for exact output of debug level messages too?
> 
> I am not sure. I think maybe not: depending on the system it would possibly be different output. 
> 

QUESTION 

On the other hand having the flags being checked would be good. Which brings up another point I referred to earlier that could possibly change the output here. 

Let's say that parsed is true but converted is false. This means that the value cannot be relied on. Instead we have to use the string. Okay. But what do we print besides the string? Can we ever rely on any of the bools being accurate? Code that I saw suggested it might be but I don't know nor do I know what should or should not be printed. 

This could change how this is done at least as far as the cases where converted is false and parsed is true. 

What about if both bools are false? Even though that should technically be an error that we won't encounter should anything be done with it anyway? If so what?

I don't know how long it will be before I can be at the laptop but hopefully I will feel like I am feeling now when I am because I want to work on this.

**Comment by xexyl** [comment]
Before I work on this I had a thought that might greatly simplify this all. I wonder what you think of this. Then I'll just hold off: that will also give me time to do other things too including hopefully get a bit of rest and feel better.

If we go the function this stuff is printed in and we look at the `JSON_STRING` case we see it much simpler:

```c
                fprint(stream, "\tlen{%s%s%s%s%s%s%s%s%s}: %ju\tvalue:\t",
                                CONVERTED_JSON_NODE(item)?"c":"",
                                PARSED_JSON_NODE(item)?"p":"",
                                item->quote ? "q" : "",
                                item->same ? "=" : "",
                                item->has_nul ? "0" : "",
                                item->slash ? "/" : "",
                                item->posix_safe ? "P" : "",
                                item->first_alphanum ? "a" : "",
                                item->upper ? "U" : "",
                                (uintmax_t)item->str_len);
                (void) fprint_line_buf(stream, item->str, item->str_len, '"', '"');

```

Now since for the number we just print it in the function as the `as_str` couldn't we just do something like this to make it so we don't have to check for all the different booleans? It would I believe make it less paths to go check and go through, it could be centralised and it might be that some code (not sure of this) would not have to be duplicated.

Can you think of a reason to not do it like strings? If not I'll go this route which I think will be much easier to follow and thus put in place.

I'll see about replying to the pending comment in the other thread and then I'm going to do other things.

**Comment by lcn2** [comment]
> # QUESTION
> 
> 
> 
> Should the test script check for exact output of debug level messages too?

We don't think so.

**Comment by lcn2** [comment]
> Let's say that parsed is true but converted is false. This means that the value cannot be relied on. Instead we have to use the string. Okay. But what do we print besides the string? Can we ever rely on any of the bools being accurate? Code that I saw suggested it might be but I don't know nor do I know what should or should not be printed

That is incorrect. If `parsed` is true, then the JSON node has been parsed and therefore the booleans and other values in the "struct json_foo" **ARE** correct: this is what parsing the node means.

If `parsed` is false, then the rest of the "struct json_foo" is unreliable (probably zeroized).

**Comment by xexyl** [comment]
> > # QUESTION
> > Should the test script check for exact output of debug level messages too?
> 
> We don't think so.

Agree.

**Comment by xexyl** [comment]
> > Let's say that parsed is true but converted is false. This means that the value cannot be relied on. Instead we have to use the string. Okay. But what do we print besides the string? Can we ever rely on any of the bools being accurate? Code that I saw suggested it might be but I don't know nor do I know what should or should not be printed
> 
> That is incorrect. If `parsed` is true, then the JSON node has been parsed and therefore the booleans and other values in the "struct json_foo" **ARE** correct: this is what parsing the node means.
> 
> If `parsed` is false, then the rest of the "struct json_foo" is unreliable (probably zeroized).

What I meant is the number variable could not be relied upon. I was not sure if the booleans can be. I thought they could be relied on and this confirms it but I wasn't sure without looking at the code.

**Comment by lcn2** [comment]
> What about if both bools are false? Even though that should technically be an error that we won't encounter should anything be done with it anyway? If so what?

Well if `parsed` is false, then the JSON node was **NOT PARSED** and so the rest of the `struct json_foo` is unreliable.  Most likely the rest of the `struct json_foo` was zeroized.  Nevertheless it is **NOT AN ERROR** for both `parsed` and `converted` to be false.  In fact it is more likely that if `parsed` is false, the `converted` will be false.

Now if `parsed` is false and `converted` is true, that also is **NOT AN ERROR**.  Again when `parsed` is false, the rest of the "struct json_foo" is unreliable, so in this case `converted` can be any value.

When `parsed` is true, then the rest of the 'struct json_foo" is valid and reflects the state of the JSON node that was parsed.  This means that `converted` can be true or false, depending on if the JSON node value was convertible.  So this JSON is valid and parsible:

```json
1e1000000000
```

So `parsed` will be true and the rest of the `struct json_number` is valid.  However because "1e1000000000" cannot be converted into a number, then `converted` will be false.  However `as_str` will still point at the JSON "value" string of "1e1000000000".

**Comment by xexyl** [comment]
> > What about if both bools are false? Even though that should technically be an error that we won't encounter should anything be done with it anyway? If so what?
> 
> Well if `parsed` is false, then the JSON node was **NOT PARSED** and so the rest of the `struct json_foo` is unreliable. Most likely the rest of the `struct json_foo` was zeroized. Nevertheless it is **NOT AN ERROR** for both `parsed` and `converted` to be false. In fact it is more likely that if `parsed` is false, the `converted` will be false.
> 
> Now if `parsed` is false and `converted` is true, that also is **NOT AN ERROR**. Again when `parsed` is false, the rest of the "struct json_foo" is unreliable, so in this case `converted` can be any value.
> 
> When `parsed` is true, then the rest of the 'struct json_foo" is valid and reflects the state of the JSON node that was parsed. This means that `converted` can be true or false, depending on if the JSON node value was convertible. So this JSON is valid and parsible:
> 
> ```json
> 1e1000000000
> ```
> 
> So `parsed` will be true and the rest of the `struct json_number` is valid. However because "1e1000000000" cannot be converted into a number, then `converted` will be false. However `as_str` will still point at the JSON "value" string of "1e1000000000".

It was an error if it wasn't converted so I made it instead an error if neither are true. Is that an error itself ? I only updated it.

I can look at the other things in more detail later - possibly not until tomorrow.

**Comment by lcn2** [comment]
> Now since for the number we just print it in the function as the as_str couldn't we just do something like this to make it so we don't have to check for all the different booleans? It would I believe make it fewer paths to check and go through, it could be centralised and it might be that some code (not sure of this) would not have to be duplicated.

It could be done that way.  In the case of `struct json_number` there are a lot of booleans, so the short boolean related letters all smashed together might be hard to read so instead of say adding "16", you might want to comma separate then an in addition ",16".

Moreover because there are so many booleans in `struct json_number`, printing them all at `-J 3` would be a bit much.  That is why you could skip some of the obvious things such as is the number fits into an 8-bit integer, you know it will fit into a 16-bit, 32-bit, 64-bit integers, etc.

There might be a good reason to see all booleans at a higher debug level such as `-J 7`.

**Comment by lcn2** [comment]
If the enhancement part of this issue seems too confusing or too unclear, let us know as we can code this debug code as needed.

**Comment by xexyl** [comment]
> If the enhancement part of this issue seems too confusing or too unclear, let us know as we can code this debug code as needed.

I don't think it is .. we'll see. Thanks.

Just taking it easy today though.

**Comment by lcn2** [comment]
> > > Let's say that parsed is true but converted is false. This means that the value cannot be relied on. Instead we have to use the string. Okay. But what do we print besides the string? Can we ever rely on any of the bools being accurate? Code that I saw suggested it might be but I don't know nor do I know what should or should not be printed
> 
> > 
> 
> > That is incorrect. If `parsed` is true, then the JSON node has been parsed and therefore the booleans and other values in the "struct json_foo" **ARE** correct: this is what parsing the node means.
> 
> > 
> 
> > If `parsed` is false, then the rest of the "struct json_foo" is unreliable (probably zeroized).
> 
> 
> 
> What I meant is the number variable could not be relied upon. I was not sure if the booleans can be. I thought they could be relied on and this confirms it but I wasn't sure without looking at the code.
> 
> 
> 
> But still what should be printed ?

When `parsed` is false (i.e., the JSON node has not yet been parsed, or was not able to be parsed, or some other error occurred that prevented JSON parsing from completing), the resulting "struct json_foo" has only one element that has meaning: that `parsed` is false.  All other elements in the "struct json_foo" have no meaning, have no reliable value and should not be used.

As for what to print: well the `as_str` pointer cannot be depended on pointing to anything valid or useful.  One could print the conventional "((NULL))" string that is sometimes used when there is an attempt to print a NULL pointer (i.e., code detects the the string to be printed "== NULL" and prints "((NULL))" instead of crashing).

When a "struct json_foo" is created by `calloc(3)`, after checking that `calloc(3)` returned a non-NULL pointer, the `parsed` value should then be explicitly be set to false.  Only after the rest of the "struct json_foo" is filled out should the `parsed` value be set to true.  This way, **UNTIL AND UNLESS** the entire rest of the "struct json_foo" is assigned values, the `parsed` value remains false in the function that is filling out the "struct json_foo".  Only at the very end of a successful filling out of the "struct json_foo" should the `parsed` value be set to true.

This way the `parsed` value acts as a sort of **LOCK** on the validity of the "struct json_foo" as it be being setup and being filled on.  Only at the last minute, just before returning a filled in "struct json_foo" should the `parsed` value be set to true.  So if one were to example a crash dump, presence of a `parsed` that is false will tell you that the "struct json_foo" was not completed, or there was a memory smash that turned `parsed` to false.

**Comment by lcn2** [comment]
> > > What about if both bools are false? Even though that should technically be an error that we won't encounter should anything be done with it anyway? If so what?
> 
> > 
> 
> > Well if `parsed` is false, then the JSON node was **NOT PARSED** and so the rest of the `struct json_foo` is unreliable. Most likely the rest of the `struct json_foo` was zeroized. Nevertheless it is **NOT AN ERROR** for both `parsed` and `converted` to be false. In fact it is more likely that if `parsed` is false, the `converted` will be false.
> 
> > 
> 
> > Now if `parsed` is false and `converted` is true, that also is **NOT AN ERROR**. Again when `parsed` is false, the rest of the "struct json_foo" is unreliable, so in this case `converted` can be any value.
> 
> > 
> 
> > When `parsed` is true, then the rest of the 'struct json_foo" is valid and reflects the state of the JSON node that was parsed. This means that `converted` can be true or false, depending on if the JSON node value was convertible. So this JSON is valid and parsible:
> 
> > 
> 
> > ```json
> 
> > 1e1000000000
> 
> > ```
> 
> > 
> 
> > So `parsed` will be true and the rest of the `struct json_number` is valid. However because "1e1000000000" cannot be converted into a number, then `converted` will be false. However `as_str` will still point at the JSON "value" string of "1e1000000000".
> 
> 
> 
> It was an error if it wasn't converted so I made it instead an error if neither are true. Is that an error itself ? I only updated it.
> 
> 
> 
> I can look at the other things in more detail later - possibly not until tomorrow.
> 
> 
> 
> ## UPDATE 0
> 
> 
> 
> See function `parse_json_number()` in json_parse.c. Other functions were updated likewise.

The purpose of issue #752 was that valid JSON such as:

```json
1e1000000000
```

Could be parsed as valid JSON even though the JSON number could not be converted into a machine value form.

So in the case `parsed` should be true, `converted` should be false, and `as_str` for the `struct json_number` would point at a string containing "1e1000000000", while all of the other "is_foo" booleans will be false because the huge value cannot bit into any of those data types.

Nevertheless in the case the JSON is valid, and `parsed` is true.  So while `converted` is false, a tool such as `jval` or `jnamval` can still use the `as_str` string to compare, print, etc.

**Comment by lcn2** [comment]
> ## UPDATE 0
> See function `parse_json_number()` in json_parse.c. Other functions were updated likewise.

See commit d9b18ee73153fa863590386eece4c5de1eaa2a16 for a paranoia change to json_alloc() to perform explicit setting of the `parsed` and `converted` booleans as well as pointers to NULL.

The comment added explains the paranoia and why these explicit sets were done.

**Comment by lcn2** [comment]
> > > What about if both bools are false? Even though that should technically be an error that we won't encounter should anything be done with it anyway? If so what?
> > 
> > 
> > Well if `parsed` is false, then the JSON node was **NOT PARSED** and so the rest of the `struct json_foo` is unreliable. Most likely the rest of the `struct json_foo` was zeroized. Nevertheless it is **NOT AN ERROR** for both `parsed` and `converted` to be false. In fact it is more likely that if `parsed` is false, the `converted` will be false.
> > Now if `parsed` is false and `converted` is true, that also is **NOT AN ERROR**. Again when `parsed` is false, the rest of the "struct json_foo" is unreliable, so in this case `converted` can be any value.
> > When `parsed` is true, then the rest of the 'struct json_foo" is valid and reflects the state of the JSON node that was parsed. This means that `converted` can be true or false, depending on if the JSON node value was convertible. So this JSON is valid and parsible:
> > ```json
> > 1e1000000000
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > So `parsed` will be true and the rest of the `struct json_number` is valid. However because "1e1000000000" cannot be converted into a number, then `converted` will be false. However `as_str` will still point at the JSON "value" string of "1e1000000000".
> 
> It was an error if it wasn't converted so I made it instead an error if neither are true. Is that an error itself ? I only updated it.
> 
> I can look at the other things in more detail later - possibly not until tomorrow.
> 
> ## UPDATE 0
> See function `parse_json_number()` in json_parse.c. Other functions were updated likewise.

We fixed the process in commit 49e2b7c63e99b96f1cbe6cc592fe32d96e852195.

**Comment by xexyl** [comment]
> > ## UPDATE 0
> > See function `parse_json_number()` in json_parse.c. Other functions were updated likewise.
> 
> See commit [d9b18ee](https://github.com/ioccc-src/mkiocccentry/commit/d9b18ee73153fa863590386eece4c5de1eaa2a16) for a paranoia change to json_alloc() to perform explicit setting of the `parsed` and `converted` booleans as well as pointers to NULL.
> 
> The comment added explains the paranoia and why these explicit sets were done.

I did: I thought it looks great. Thanks!

**Comment by xexyl** [comment]
> > > > What about if both bools are false? Even though that should technically be an error that we won't encounter should anything be done with it anyway? If so what?
> > > 
> > > 
> > > Well if `parsed` is false, then the JSON node was **NOT PARSED** and so the rest of the `struct json_foo` is unreliable. Most likely the rest of the `struct json_foo` was zeroized. Nevertheless it is **NOT AN ERROR** for both `parsed` and `converted` to be false. In fact it is more likely that if `parsed` is false, the `converted` will be false.
> > > Now if `parsed` is false and `converted` is true, that also is **NOT AN ERROR**. Again when `parsed` is false, the rest of the "struct json_foo" is unreliable, so in this case `converted` can be any value.
> > > When `parsed` is true, then the rest of the 'struct json_foo" is valid and reflects the state of the JSON node that was parsed. This means that `converted` can be true or false, depending on if the JSON node value was convertible. So this JSON is valid and parsible:
> > > ```json
> > > 1e1000000000
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > So `parsed` will be true and the rest of the `struct json_number` is valid. However because "1e1000000000" cannot be converted into a number, then `converted` will be false. However `as_str` will still point at the JSON "value" string of "1e1000000000".
> > 
> > 
> > It was an error if it wasn't converted so I made it instead an error if neither are true. Is that an error itself ? I only updated it.
> > I can look at the other things in more detail later - possibly not until tomorrow.
> > ## UPDATE 0
> > See function `parse_json_number()` in json_parse.c. Other functions were updated likewise.
> 
> We fixed the process in commit [49e2b7c](https://github.com/ioccc-src/mkiocccentry/commit/49e2b7c63e99b96f1cbe6cc592fe32d96e852195).

I saw: looks great. Thanks!

**Comment by xexyl** [comment]
> > What about if both bools are false? Even though that should technically be an error that we won't encounter should anything be done with it anyway? If so what?
> 
> Well if `parsed` is false, then the JSON node was **NOT PARSED** and so the rest of the `struct json_foo` is unreliable. Most likely the rest of the `struct json_foo` was zeroized. Nevertheless it is **NOT AN ERROR** for both `parsed` and `converted` to be false. In fact it is more likely that if `parsed` is false, the `converted` will be false.

That does make sense. However when I added the macros I just updated the check. Since it had checked for `!converted` and that was an ERROR, I made it so that if both `!converted` and `!parsed` it is an error. It sounds like that itself is in error but if it's not parsed wouldn't it indicate that the data is invalid and thus the json is invalid?

> 
> Now if `parsed` is false and `converted` is true, that also is **NOT AN ERROR**. Again when `parsed` is false, the rest of the "struct json_foo" is unreliable, so in this case `converted` can be any value.
> 
Hmm ... this seems more like a bogus case? I thought `parsed` means it could be parsed but `converted` means it can actually be represented in C: especially useful for numbers.


> When `parsed` is true, then the rest of the 'struct json_foo" is valid and reflects the state of the JSON node that was parsed. This means that `converted` can be true or false, depending on if the JSON node value was convertible. So this JSON is valid and parsible:

I have it checked - to see if it's valid - both parsed and converted are true OR just parsed is true. What should it be then? And what about the fact I only updated the code to check for one or the other (or both) and if both are false it's an error? For instance in function `parse_json_number()`:

```c
    item = &(number->item.number);
    if (!VALID_JSON_NODE(item)) {
        err(170, __func__, "couldn't convert number string: <%s>", string);
        not_reached();
    }
```

I just updated it from checking `!item->converted`. What _should_ it be if it's not supposed to be an error? It sounds like I should remove this check and as part of this issue. I'll do that: esp as the boolean in print status suggests that the booleans should both be printed depending on the debug level. If however this is incorrect (it does not seem like it but if there's a misunderstanding) I will then close the pull request or make any corrections necessary. Again since this seems important to the this issue I'll do it first: maybe after I reply to the comments here.



> 
> ```json
> 1e1000000000
> ```
> 
> So `parsed` will be true and the rest of the `struct json_number` is valid. However because "1e1000000000" cannot be converted into a number, then `converted` will be false. However `as_str` will still point at the JSON "value" string of "1e1000000000".


Yep. That makes sense to me.

**Comment by xexyl** [comment]
> > Now since for the number we just print it in the function as the as_str couldn't we just do something like this to make it so we don't have to check for all the different booleans? It would I believe make it fewer paths to check and go through, it could be centralised and it might be that some code (not sure of this) would not have to be duplicated.
> 
> It could be done that way. In the case of `struct json_number` there are a lot of booleans, so the short boolean related letters all smashed together might be hard to read so instead of say adding "16", you might want to comma separate then an in addition ",16".

I agree it should be comma separated. What I meant is that rather than have all the `if (..)` checks within other if checks have it just print the debug statements based on debug level and using the ternary operator. That would be simpler to follow the code and also simplify the writing of the code. I'll do that.

> 
> Moreover because there are so many booleans in `struct json_number`, printing them all at `-J 3` would be a bit much. That is why you could skip some of the obvious things such as is the number fits into an 8-bit integer, you know it will fit into a 16-bit, 32-bit, 64-bit integers, etc.
> 
Of course.

> There might be a good reason to see all booleans at a higher debug level such as `-J 7`.

Sure.

**Comment by xexyl** [comment]
> > > > Let's say that parsed is true but converted is false. This means that the value cannot be relied on. Instead we have to use the string. Okay. But what do we print besides the string? Can we ever rely on any of the bools being accurate? Code that I saw suggested it might be but I don't know nor do I know what should or should not be printed
> > 
> > 
> > > 
> > 
> > 
> > > That is incorrect. If `parsed` is true, then the JSON node has been parsed and therefore the booleans and other values in the "struct json_foo" **ARE** correct: this is what parsing the node means.
> > 
> > 
> > > 
> > 
> > 
> > > If `parsed` is false, then the rest of the "struct json_foo" is unreliable (probably zeroized).
> > 
> > 
> > What I meant is the number variable could not be relied upon. I was not sure if the booleans can be. I thought they could be relied on and this confirms it but I wasn't sure without looking at the code.
> > But still what should be printed ?
> 
> When `parsed` is false (i.e., the JSON node has not yet been parsed, or was not able to be parsed, or some other error occurred that prevented JSON parsing from completing), the resulting "struct json_foo" has only one element that has meaning: that `parsed` is false. All other elements in the "struct json_foo" have no meaning, have no reliable value and should not be used.

Now it sounds like it IS an error if it cannot be parsed as you wrote 'or was not able to be parsed, or some other _ERROR_ occurred that prevented JSON parsing from completing ...'. So I'm not sure what it should be now. It might be that you mean if parsed is false then it's not an error but there might be some error condition. But in the case of error isn't it fatal so the checks of the bools won't happen anyway?



> 
> As for what to print: well the `as_str` pointer cannot be depended on pointing to anything valid or useful. One could print the conventional "((NULL))" string that is sometimes used when there is an attempt to print a NULL pointer (i.e., code detects the the string to be printed "== NULL" and prints "((NULL))" instead of crashing).

I like the idea of printing that yes.

> 
> When a "struct json_foo" is created by `calloc(3)`, after checking that `calloc(3)` returned a non-NULL pointer, the `parsed` value should then be explicitly be set to false. 

As it is now and was.

> Only after the rest of the "struct json_foo" is filled out should the `parsed` value be set to true. This way, **UNTIL AND UNLESS** the entire rest of the "struct json_foo" is assigned values, the `parsed` value remains false in the function that is creating and filling out the "struct json_foo". Only at the very end of a successful filling out of the "struct json_foo" should the `parsed` value be set to true by the function that is creating and filling out the "struct json_foo".

Yes. That's how I did it I thought? Unless I made an error like what prompted this issue to be opened (and which I fixed)?

> 
> This way the `parsed` value acts as a sort of **LOCK** on the validity of the "struct json_foo" as it be being setup and being filled on. Only at the last minute, just before returning a filled in "struct json_foo" should the `parsed` value be set to true. So if one were to example a crash dump, presence of a `parsed` that is false will tell you that the "struct json_foo" was not completed, or there was a memory smash that turned `parsed` to false.

Yes. But it sounds like errors are not fatal now and that it should just print `converted` and `parsed` to false. Since one should always check that the node is valid before use it might not be a problem though some might argue that if both are false it could be fatal. Not sure but since the regexp in the lexer should not return anything that's invalid (except maybe strings which we check explicitly) this isn't an issue so much?

**Comment by xexyl** [comment]
> > > > What about if both bools are false? Even though that should technically be an error that we won't encounter should anything be done with it anyway? If so what?
> > 
> > 
> > > 
> > 
> > 
> > > Well if `parsed` is false, then the JSON node was **NOT PARSED** and so the rest of the `struct json_foo` is unreliable. Most likely the rest of the `struct json_foo` was zeroized. Nevertheless it is **NOT AN ERROR** for both `parsed` and `converted` to be false. In fact it is more likely that if `parsed` is false, the `converted` will be false.
> > 
> > 
> > > 
> > 
> > 
> > > Now if `parsed` is false and `converted` is true, that also is **NOT AN ERROR**. Again when `parsed` is false, the rest of the "struct json_foo" is unreliable, so in this case `converted` can be any value.
> > 
> > 
> > > 
> > 
> > 
> > > When `parsed` is true, then the rest of the 'struct json_foo" is valid and reflects the state of the JSON node that was parsed. This means that `converted` can be true or false, depending on if the JSON node value was convertible. So this JSON is valid and parsible:
> > 
> > 
> > > 
> > 
> > 
> > > 
> > 
> > 
> > > 1e1000000000
> > 
> > 
> > > 
> > 
> > 
> > > 
> > 
> > 
> > > So `parsed` will be true and the rest of the `struct json_number` is valid. However because "1e1000000000" cannot be converted into a number, then `converted` will be false. However `as_str` will still point at the JSON "value" string of "1e1000000000".
> > 
> > 
> > It was an error if it wasn't converted so I made it instead an error if neither are true. Is that an error itself ? I only updated it.
> > I can look at the other things in more detail later - possibly not until tomorrow.
> > ## UPDATE 0
> > See function `parse_json_number()` in json_parse.c. Other functions were updated likewise.
> 
> The purpose of issue #752 was that valid JSON such as:
> 
> ```json
> 1e1000000000
> ```
> 
> Could be parsed as valid JSON even though the JSON number could not be converted into a machine value form.
> 
That's what I did yes.

> So in the case `parsed` should be true, `converted` should be false, and `as_str` for the `struct json_number` would point at a string containing "1e1000000000", while all of the other "is_foo" booleans will be false because the huge value cannot bit into any of those data types.

Yes: done in that issue resolution.

> 
> Nevertheless in the case the JSON is valid, and `parsed` is true. So while `converted` is false, a tool such as `jval` or `jnamval` can still use the `as_str` string to compare, print, etc.

True.

**Comment by xexyl** [comment]
With commit https://github.com/ioccc-src/mkiocccentry/pull/792/commits/454910f4d35af60de48e0c7ba3e4ac26813fea35 the condition of `VALID_JSON_NODE` in the parse functions is no longer an error.

I'm going to try and rest a bit and when I'm back I hope to look at this issue a bit more since it seems that it has higher priority than the other one and in fact the other issue is dependent on this one being resolved.

I should say that this afternoon I should have more time at the laptop so hopefully we can get more done today: but we'll see on that .. depends on how I feel at that point! I very possibly will feel the desire to watch a film today too but I'd also like to resolve this issue and get moving on the other one.

Well rest time now. I hope that I can get back to sleep but I should be able to rest a while if not sleep - so it feels like at this time anyway.

Good night! (even though it's technically morning for 3 hours and 20 minutes and 39 seconds though now already more seconds! :-)

**Comment by xexyl** [comment]
FYI: `make test` fails. I guess with one of the fixes you made. I'll address that when I'm back. Time to rest.

**Comment by xexyl** [comment]
.. oh it's very simple. Two files are no longer considered bad? I'll fix that very quickly and then go rest.

**Comment by lcn2** [comment]
Well we can see that there is question of what it means for there to be an error.

Let's use the them "there is an error" to mean that something i seriously wrong and that a function needs to abort: perhaps by calling `err()` to an equivalent halting function OR for functions that must return, they do so without completing their intended action.

An example of "there is an error" is a firewall code block early on in a function where some pointer to a structure is found to be NULL.  The early firewall code in that function, if it detects the pointer to a function is NULL (and thus later on the function would SEGFAULT if it did something like ptr->stuff where ptr == NULL) then "there is an error".  The function has to either call `err()` (or related exiting function) and die **OR** the function as to return early and presumably return something that tells the calling function that there was a failure.

What is an example of "there is NOT an error"?  Well being given a JSON number "1e1000000000" that while the number cannot be converted, the JSON node is parsed.  So in this case `parsed` is true while `converted` is false.  The presence of such a NODE "is not an error".

Now later on, if there is some semantical check, a JSON number could be an error.  Say the JSON number "1e1000000000" happens to be the entry number.  The check to be sure that the entry number is ">= 0" and "<= MAX_ENTRY_NUM".  In this case the fact that `converted` is false "IS AN ERROR" because there is no entry number integer value that can be checked.

Sure, one could try to go through the effort to see that the JSON node was a valid parsed JSON number, and them try to make sense of the huge value in the string .. but that is a bit silly.  :-)

**Comment by lcn2** [comment]
Regarding pull request #792:

```
#define VALID_JSON_NODE(item) ((item)->converted || (item)->parsed)
#define CONVERTED_PARSED_JSON_NODE(item) ((item)->converted && (item)->parsed)
#define CONVERTED_JSON_NODE(item) ((item)->converted && !(item)->parsed)
#define PARSED_JSON_NODE(item) ((item)->parsed && !(item)->converted)
```

If the JSON parser discovers that a JSON node, such as a JSON number, after calling `json_conv_number()`, finds that the "item->parsed" is false, then the JSON parser has discovered that the JSON document is NOT valid JSON.  Is that "an error"?  Well one certainly would not want the parser to call `err()` and die, so in that sense NO.  On the other hand the JSON parser should NOT report that the JSON is valid either.  Instead the call to `parse_json(ptr, len, &flag)` should cause the `*flag` to be set to `false`.

**Comment by xexyl** [comment]
> Regarding pull request #792:
> 
> If the JSON parser discovers that a JSON node, such as a JSON number, after calling `json_conv_number()`, finds that the "item->parsed" is false, then the JSON parser has discovered that the JSON document is NOT valid JSON. Is that "an error"? Well one certainly would not want the parser to call `err()` and die, so in that sense NO. On the other hand the JSON parser should NOT report that the JSON is valid either. Instead the call to `parse_json(ptr, len, &flag)` should cause the `*flag` to be set to `false`.
> 
Yes. I realised what I left out thought that's a better place for it. I'll be updating the code and then will make it not a draft.

> ## UPDATE 0
> We are writing this comment now, so we will soon talk about these macros that were suggested in the now closed pull request #792:
> 
Oh it's not closed: I just converted it to a draft as I noticed a mistake: something you noted above in fact. I'll correct soon.

> ```c
> #define VALID_JSON_NODE(item) ((item)->converted || (item)->parsed)
> #define CONVERTED_PARSED_JSON_NODE(item) ((item)->converted && (item)->parsed)
> #define CONVERTED_JSON_NODE(item) ((item)->converted && !(item)->parsed)
> #define PARSED_JSON_NODE(item) ((item)->parsed && !(item)->converted)
> ```
> 
> The `PARSED_JSON_NODE(item)` macro purely depends on `(item)->parsed` alone. If `parsed` is true, then the item was parsed. If `parsed` is false, then the item was NOT parsed. So this seems correct:

Correct except I realised that they should all check that `item != NULL` first. I'll be fixing this too.

> 
> ```c
> #define PARSED_JSON_NODE(item) ((item)->parsed)
> ```
> 
This could be an improvement yes. The point was originally that it could be parsed but couldn't be connected. This is actually used in code.


> What about `CONVERTED_JSON_NODE(item)`. Well that will depend on `(item)->converted` ... HOWEVER ... one can only properly test this condition IF AND ONLY IF the JSON node was parsed.
> 
> So this seems connect:
> 
> ```c
> #define CONVERTED_JSON_NODE(item) (PARSED_JSON_NODE(item) && (item)->converted)
> ```

Good call. I fixed that. Not committed yet.


> 
> We like the use of the macro in that form as it makes it clear that the JSON node must be PARSED and them the converted check happens).
> 
Agree.

> We are not sure if the `CONVERTED_PARSED_JSON_NODE(item)` is a useful macro. We recommend against the use of that macro.

Well it's useful as it shows that both are true.

> 
> However if there is some compelling reason for the `CONVERTED_PARSED_JSON_NODE(item)`, then let it be the same as the `CONVERTED_JSON_NODE(item)` macro:
> 
> ```c
> /* we recommend not using this macro */
> #define CONVERTED_PARSED_JSON_NODE(item) (CONVERTED_JSON_NODE(item))
> ```
> 
I could do that I guess: perhaps if converted is true parsed will always be true too. I have to see what is used where. .. and it' used a lot for the reason I noted: to make sure that both are true as an extra sanity check. Is that not desired?


> The final question is what does the `VALID_JSON_NODE(item)` mean? Here we are not sure of the intent of this macro.
> 
> Consider the JSON node for the JSON number "1e1000000000", where `parsed` is true but `converted` is false. Is that node valid? It is a valid JSON syntax so in one sense it is valid. But the numeric value cannot be converted into a machine integer value, so does that make the node invalid? From the perspective of the so-called JSON spec, no: the JSON number "1e1000000000" is a valid JSON number. From the perspective of code that want to do an integer bounds check, the node is not useful.
> 

It means exactly that. It couldn't be converted but it is valid JSON. True that the bounds check is false and it's not useful. But it can be used to detect if the json is valid. It'll do so in a better place - a place you mentioned.

**Comment by xexyl** [comment]
> > Regarding pull request #792:
> > If the JSON parser discovers that a JSON node, such as a JSON number, after calling `json_conv_number()`, finds that the "item->parsed" is false, then the JSON parser has discovered that the JSON document is NOT valid JSON. Is that "an error"? Well one certainly would not want the parser to call `err()` and die, so in that sense NO. On the other hand the JSON parser should NOT report that the JSON is valid either. Instead the call to `parse_json(ptr, len, &flag)` should cause the `*flag` to be set to `false`.
> 
> Yes. I realised what I left out thought that's a better place for it. I'll be updating the code and then will make it not a draft.
> 

Actually I don't think it can be done here: not directly. Because those booleans are per node not the json itself. So there has to be a way to pass that information to the overall parse function. That might be why the above files triggered valid JSON when it's not technically valid JSON. The fact you don't want it to be fatal causes this problem. The files I found being a problem are NOT valid JSON and so should be in the bad/ directory but removing the check for a valid json node caused the problem.

**Comment by xexyl** [comment]
> > > Regarding pull request #792:
> > > If the JSON parser discovers that a JSON node, such as a JSON number, after calling `json_conv_number()`, finds that the "item->parsed" is false, then the JSON parser has discovered that the JSON document is NOT valid JSON. Is that "an error"? Well one certainly would not want the parser to call `err()` and die, so in that sense NO. On the other hand the JSON parser should NOT report that the JSON is valid either. Instead the call to `parse_json(ptr, len, &flag)` should cause the `*flag` to be set to `false`.
> > 
> > 
> > Yes. I realised what I left out thought that's a better place for it. I'll be updating the code and then will make it not a draft.
> 
> Actually I don't think it can be done here: not directly. Because those booleans are per node not the json itself. So there has to be a way to pass that information to the overall parse function. That might be why the above files triggered valid JSON when it's not technically valid JSON. The fact you don't want it to be fatal causes this problem. The files I found being a problem are NOT valid JSON and so should be in the bad/ directory but removing the check for a valid json node caused the problem.


## UPDATE 0

Tried to add booleans to the top node but that won't work either: the parse_ ... functions inn the parser (`jparse.y`) would have to be greatly complicated to do this. What do you propose is done? By removing the check for either being valid it breaks the parser.

I'll close the pull request now as I don't think it's correct and trying to fix it this way is problematic.

**Comment by lcn2** [comment]
> I could do that I guess: perhaps if converted is true parsed will always be true too. I have to see what is used where. .. and it' used a lot for the reason I noted: to make sure that both are true as an extra sanity check. Is that not desired?

One can only use a true `converted` value **IF AND ONLY IF** the `parsed` is true.

If `parsed` is false, then `converted` has no meaning.

**Comment by xexyl** [comment]
A recent change has broken `make test`. The parser now thinks that two files are valid. I wonder if it's commit d9b18ee73153fa863590386eece4c5de1eaa2a16. I haven't tried it yet but I'm going to now.

**Comment by xexyl** [comment]
> A recent change has broken `make test`. The parser now thinks that two files are valid. I wonder if it's commit [d9b18ee](https://github.com/ioccc-src/mkiocccentry/commit/d9b18ee73153fa863590386eece4c5de1eaa2a16). I haven't tried it yet but I'm going to now.

Okay for some reason file were moved from bad to good .. not sure how or why. Changing them back and the log looks okay and it works. Given his happened and I don't known how I think I better not do anything right now. I'll improve those macros later today - I hope.

**Comment by xexyl** [comment]
> > I could do that I guess: perhaps if converted is true parsed will always be true too. I have to see what is used where. .. and it' used a lot for the reason I noted: to make sure that both are true as an extra sanity check. Is that not desired?
> 
> One can only use a true `converted` value **IF AND ONLY IF** the `parsed` is true.
> 
> If `parsed` is false, then `converted` has no meaning.

Right. Well I'll improve the macro as suggested later on ... hopefully today. I'm not feeling so well right now. I hope it passes. It might very well but I need a break now anyway.

**Comment by xexyl** [comment]
> > > Regarding pull request #792:
> > > If the JSON parser discovers that a JSON node, such as a JSON number, after calling `json_conv_number()`, finds that the "item->parsed" is false, then the JSON parser has discovered that the JSON document is NOT valid JSON. Is that "an error"? Well one certainly would not want the parser to call `err()` and die, so in that sense NO. On the other hand the JSON parser should NOT report that the JSON is valid either. Instead the call to `parse_json(ptr, len, &flag)` should cause the `*flag` to be set to `false`.
> > 
> > 
> > Yes. I realised what I left out thought that's a better place for it. I'll be updating the code and then will make it not a draft.
> 
> Actually I don't think it can be done here: not directly. Because those booleans are per node not the json itself. So there has to be a way to pass that information to the overall parse function. That might be why the above files triggered valid JSON when it's not technically valid JSON. The fact you don't want it to be fatal causes this problem. The files I found being a problem are NOT valid JSON and so should be in the bad/ directory but removing the check for a valid json node caused the problem.

I don't think this is a problem after all: at least not as far as the make test failing. That is now resolved. But as the change was not strictly correct I don't know what to do either.

I'm going to take a break now .. hope to be back later.

**Comment by xexyl** [comment]
> Regarding pull request #792:
> 
> If the JSON parser discovers that a JSON node, such as a JSON number, after calling `json_conv_number()`, finds that the "item->parsed" is false, then the JSON parser has discovered that the JSON document is NOT valid JSON. Is that "an error"? Well one certainly would not want the parser to call `err()` and die, so in that sense NO. On the other hand the JSON parser should NOT report that the JSON is valid either. Instead the call to `parse_json(ptr, len, &flag)` should cause the `*flag` to be set to `false`.
> 
> ## UPDATE 0
> Regarding the macros that were suggested in the (now closed) pull request #792:
> 
> ```c
> #define VALID_JSON_NODE(item) ((item)->converted || (item)->parsed)
> #define CONVERTED_PARSED_JSON_NODE(item) ((item)->converted && (item)->parsed)
> #define CONVERTED_JSON_NODE(item) ((item)->converted && !(item)->parsed)
> #define PARSED_JSON_NODE(item) ((item)->parsed && !(item)->converted)
> ```
> 
> The `PARSED_JSON_NODE(item)` macro purely depends on `(item)->parsed` alone. If `parsed` is true, then the item was parsed. If `parsed` is false, then the item was NOT parsed. So this seems correct:
> 
> ```c
> #define PARSED_JSON_NODE(item) ((item)->parsed)
> ```
> 
> What about the `CONVERTED_JSON_NODE(item)` macro? Well that will depend on `(item)->converted`. **HOWEVER** one can only properly test this condition IF AND ONLY IF the JSON node was parsed.
> 
> So this seems connect:
> 
> ```c
> #define CONVERTED_JSON_NODE(item) (PARSED_JSON_NODE(item) && (item)->converted)
> ```
> 
> We like the use of the macro in that form as it makes it clear that the JSON node must be PARSED and them the converted check happens).
> 
> We are not sure if the `CONVERTED_PARSED_JSON_NODE(item)` is a useful macro. We recommend against the use of that macro.
> 
> However if there is some compelling reason for the `CONVERTED_PARSED_JSON_NODE(item)`, then let it be the same as the `CONVERTED_JSON_NODE(item)` macro:
> 
> ```c
> /* we recommend not using this macro */
> #define CONVERTED_PARSED_JSON_NODE(item) (CONVERTED_JSON_NODE(item))
> ```
> 
> The final question is what does the `VALID_JSON_NODE(item)` mean? Here we are not sure of the intent of this macro.
> 
> Consider the JSON node for the JSON number "1e1000000000", where `parsed` is true but `converted` is false. Is that node valid? It is a valid JSON syntax so in one sense it is valid. But the numeric value cannot be converted into a machine integer value, so does that make the node invalid? From the perspective of the so-called JSON spec, no: the JSON number "1e1000000000" is a valid JSON number. From the perspective of code that want to do an integer bounds check, the node is not useful.
> 
> The problem we have with the `VALID_JSON_NODE(item)` macro concept is that what is "VALID" is context dependent.
> 
> ## UPDATE 1
> What macro should `sem_node_valid()` use? One could argue that the use in that function of the macro `VALID_JSON_NODE(item)` would be better if it was the `PARSED_JSON_NODE(item)`.
> 
> So do we want the `VALID_JSON_NODE(item)` macro? Maybe not.
> 
> Maybe all that is needed is:
> 
> ```c
> #define PARSED_JSON_NODE(item) ((item)->parsed)
> #define CONVERTED_JSON_NODE(item) (PARSED_JSON_NODE(item) && (item)->converted)
> ```
> 
> ???

I'll address the updates sometime later.

**Comment by xexyl** [comment]
> Well we can see that there is question of what it means for there to be an error.
> 
> Let's use the them "there is an error" to mean that something i seriously wrong and that a function needs to abort: perhaps by calling `err()` to an equivalent halting function OR for functions that must return, they do so without completing their intended action.
> 
> An example of "there is an error" is a firewall code block early on in a function where some pointer to a structure is found to be NULL. The early firewall code in that function, if it detects the pointer to a function is NULL (and thus later on the function would SEGFAULT if it did something like ptr->stuff where ptr == NULL) then "there is an error". The function has to either call `err()` (or related exiting function) and die **OR** the function as to return early and presumably return something that tells the calling function that there was a failure.
> 
> What is an example of "there is NOT an error"? Well being given a JSON number "1e1000000000" that while the number cannot be converted, the JSON node is parsed. So in this case `parsed` is true while `converted` is false. The presence of such a NODE "is not an error".
> 
> Now later on, if there is some semantical check, a JSON number could be an error. Say the JSON number "1e1000000000" happens to be the entry number. The check to be sure that the entry number is ">= 0" and "<= MAX_ENTRY_NUM". In this case the fact that `converted` is false "IS AN ERROR" because there is no entry number integer value that can be checked.
> 
> Sure, one could try to go through the effort to see that the JSON node was a valid parsed JSON number, and them try to make sense of the huge value in the string .. but that is a bit silly. :-)
> 
> We apologize for the confusion about what it means for "there is an error" or "it is not an error".

I'll also look at this comment again later. Going to take a break now.

**Comment by lcn2** [comment]
> > > > Regarding pull request #792:
> 
> > > > If the JSON parser discovers that a JSON node, such as a JSON number, after calling `json_conv_number()`, finds that the "item->parsed" is false, then the JSON parser has discovered that the JSON document is NOT valid JSON. Is that "an error"? Well one certainly would not want the parser to call `err()` and die, so in that sense NO. On the other hand the JSON parser should NOT report that the JSON is valid either. Instead the call to `parse_json(ptr, len, &flag)` should cause the `*flag` to be set to `false`.
> 
> > > 
> 
> > > 
> 
> > > Yes. I realised what I left out thought that's a better place for it. I'll be updating the code and then will make it not a draft.
> 
> > 
> 
> > Actually I don't think it can be done here: not directly. Because those booleans are per node not the json itself. So there has to be a way to pass that information to the overall parse function. That might be why the above files triggered valid JSON when it's not technically valid JSON. The fact you don't want it to be fatal causes this problem. The files I found being a problem are NOT valid JSON and so should be in the bad/ directory but removing the check for a valid json node caused the problem.
> 
> 
> 
> I don't think this is a problem after all: at least not as far as the make test failing. That is now resolved. But as the change was not strictly correct I don't know what to do either.
> 
> 
> 
> I'm going to take a break now .. hope to be back later.

What does the parser to if the function:

```c
struct json *json_conv_number(char const *ptr, size_t len)
```

returns NULL?

What does the parser do if that same function returns a valid `struct json *` pointer BUT the `parsed` boolean is false?

There MUST/SHOULD/NEEDS be code somewhere in the parserthat checks if the "json_conv_foo()" returns a NULL pointer.  Such a code check could also call the `PARSED_JSON_NODE(item)` macro as well.

**Comment by xexyl** [comment]
Here's what happens when we remove the check for `!converted && !parsed`:

```
jparse/test_jparse/jparse_test.sh -J 0 -d ./test_ioccc/test_JSON -s general.json -j jparse/jparse jparse/test_jparse/json_teststr.txt
The following files failed: 
--

	    FILE: ./test_ioccc/test_JSON/general.json/bad/info.string.4.json
	    FILE: ./test_ioccc/test_JSON/general.json/bad/invalid.NUL.string.json
--
./ioccc_test.sh: ERROR: jparse/test_jparse/jparse_test.sh for general.json non-zero exit code: 1

EXIT_CODE set to: 28

FAILED: jparse/test_jparse/jparse_test.sh for general.json

```

because the code seems to think it's valid. Why? For the reason I noted earlier: there's no way to tell the parser that it's invalid. Which is why you, I guess, originally had the check that converted is true. I just modified it to check for converted or parsed.

And it sounds like you have a problem more with the name of `VALID_JSON_NODE` than anything else as you refer to what is considered _VALID_. But as you can see this macro IS useful.

**Comment by xexyl** [comment]
> > > > > Regarding pull request #792:
> > 
> > 
> > > > > If the JSON parser discovers that a JSON node, such as a JSON number, after calling `json_conv_number()`, finds that the "item->parsed" is false, then the JSON parser has discovered that the JSON document is NOT valid JSON. Is that "an error"? Well one certainly would not want the parser to call `err()` and die, so in that sense NO. On the other hand the JSON parser should NOT report that the JSON is valid either. Instead the call to `parse_json(ptr, len, &flag)` should cause the `*flag` to be set to `false`.
> > 
> > 
> > > >
> > 
> > 
> > > >
> > 
> > 
> > > > Yes. I realised what I left out thought that's a better place for it. I'll be updating the code and then will make it not a draft.
> > 
> > 
> > > 
> > 
> > 
> > > Actually I don't think it can be done here: not directly. Because those booleans are per node not the json itself. So there has to be a way to pass that information to the overall parse function. That might be why the above files triggered valid JSON when it's not technically valid JSON. The fact you don't want it to be fatal causes this problem. The files I found being a problem are NOT valid JSON and so should be in the bad/ directory but removing the check for a valid json node caused the problem.
> > 
> > 
> > I don't think this is a problem after all: at least not as far as the make test failing. That is now resolved. But as the change was not strictly correct I don't know what to do either.
> > I'm going to take a break now .. hope to be back later.
> 
> What does the JSON parser do if the function:
> 
> ```c
> struct json *json_conv_number(char const *ptr, size_t len)
> ```
> 
> returns NULL?
> 
> There MUST/SHOULD/NEEDS be code somewhere in the parser that checks if the "json_conv_foo()" call returns a NULL pointer. Such a code check could also call the `PARSED_JSON_NODE(item)` macro as well and treat that situation also as a "this is invalid JSON" condition.

The function never returns on NULL so that shouldn't happen?

**Comment by xexyl** [comment]
> > Regarding pull request #792:
> > If the JSON parser discovers that a JSON node, such as a JSON number, after calling `json_conv_number()`, finds that the "item->parsed" is false, then the JSON parser has discovered that the JSON document is NOT valid JSON. Is that "an error"? Well one certainly would not want the parser to call `err()` and die, so in that sense NO. On the other hand the JSON parser should NOT report that the JSON is valid either. Instead the call to `parse_json(ptr, len, &flag)` should cause the `*flag` to be set to `false`.
> > ## UPDATE 0
> > Regarding the macros that were suggested in the (now closed) pull request #792:
> > ```c
> > #define VALID_JSON_NODE(item) ((item)->converted || (item)->parsed)
> > #define CONVERTED_PARSED_JSON_NODE(item) ((item)->converted && (item)->parsed)
> > #define CONVERTED_JSON_NODE(item) ((item)->converted && !(item)->parsed)
> > #define PARSED_JSON_NODE(item) ((item)->parsed && !(item)->converted)
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > The `PARSED_JSON_NODE(item)` macro purely depends on `(item)->parsed` alone. If `parsed` is true, then the item was parsed. If `parsed` is false, then the item was NOT parsed. So this seems correct:
> > ```c
> > #define PARSED_JSON_NODE(item) ((item)->parsed)
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > What about the `CONVERTED_JSON_NODE(item)` macro? Well that will depend on `(item)->converted`. **HOWEVER** one can only properly test this condition IF AND ONLY IF the JSON node was parsed.
> > So this seems connect:
> > ```c
> > #define CONVERTED_JSON_NODE(item) (PARSED_JSON_NODE(item) && (item)->converted)
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > We like the use of the macro in that form as it makes it clear that the JSON node must be PARSED and them the converted check happens).
> > We are not sure if the `CONVERTED_PARSED_JSON_NODE(item)` is a useful macro. We recommend against the use of that macro.
> > However if there is some compelling reason for the `CONVERTED_PARSED_JSON_NODE(item)`, then let it be the same as the `CONVERTED_JSON_NODE(item)` macro:
> > ```c
> > /* we recommend not using this macro */
> > #define CONVERTED_PARSED_JSON_NODE(item) (CONVERTED_JSON_NODE(item))
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > The final question is what does the `VALID_JSON_NODE(item)` mean? Here we are not sure of the intent of this macro.
> > Consider the JSON node for the JSON number "1e1000000000", where `parsed` is true but `converted` is false. Is that node valid? It is a valid JSON syntax so in one sense it is valid. But the numeric value cannot be converted into a machine integer value, so does that make the node invalid? From the perspective of the so-called JSON spec, no: the JSON number "1e1000000000" is a valid JSON number. From the perspective of code that want to do an integer bounds check, the node is not useful.
> > The problem we have with the `VALID_JSON_NODE(item)` macro concept is that what is "VALID" is context dependent.
> > ## UPDATE 1
> > What macro should `sem_node_valid()` use? One could argue that the use in that function of the macro `VALID_JSON_NODE(item)` would be better if it was the `PARSED_JSON_NODE(item)`.
> > So do we want the `VALID_JSON_NODE(item)` macro? Maybe not.
> > Maybe all that is needed is:
> > ```c
> > #define PARSED_JSON_NODE(item) ((item)->parsed)
> > #define CONVERTED_JSON_NODE(item) (PARSED_JSON_NODE(item) && (item)->converted)
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > ???
> 
> I'll address the updates sometime later.

See my above comment where I note how one at least is needed.

**Comment by xexyl** [comment]
> > Well we can see that there is question of what it means for there to be an error.
> > Let's use the them "there is an error" to mean that something i seriously wrong and that a function needs to abort: perhaps by calling `err()` to an equivalent halting function OR for functions that must return, they do so without completing their intended action.
> > An example of "there is an error" is a firewall code block early on in a function where some pointer to a structure is found to be NULL. The early firewall code in that function, if it detects the pointer to a function is NULL (and thus later on the function would SEGFAULT if it did something like ptr->stuff where ptr == NULL) then "there is an error". The function has to either call `err()` (or related exiting function) and die **OR** the function as to return early and presumably return something that tells the calling function that there was a failure.
> > What is an example of "there is NOT an error"? Well being given a JSON number "1e1000000000" that while the number cannot be converted, the JSON node is parsed. So in this case `parsed` is true while `converted` is false. The presence of such a NODE "is not an error".
> > Now later on, if there is some semantical check, a JSON number could be an error. Say the JSON number "1e1000000000" happens to be the entry number. The check to be sure that the entry number is ">= 0" and "<= MAX_ENTRY_NUM". In this case the fact that `converted` is false "IS AN ERROR" because there is no entry number integer value that can be checked.
> > Sure, one could try to go through the effort to see that the JSON node was a valid parsed JSON number, and them try to make sense of the huge value in the string .. but that is a bit silly. :-)
> > We apologize for the confusion about what it means for "there is an error" or "it is not an error".
> 
> I'll also look at this comment again later. Going to take a break now.

By not checking that either converted or true is parsed a node can be considered valid JSON even when it's not. That will break `make test` and it will report that something that is invalid JSON is valid. That's why we need to make it a fatal error if both are false. Unless you have a way to pass it to the top level node that doesn't unnecessarily complicate it.

But as noted before I only updated the code to add the check for parsed boolean. This was done through the macro `VALID_JSON_NODE` which checks that either is true. It used to be that those functions only checked if `!converted` and that was an error.

**Comment by lcn2** [comment]
We will address the 👀 marked comments after we sleep and wake up.

**Comment by xexyl** [comment]
> We will address the 👀 marked comments after we sleep and wake up. 

Hope you got a good sleep!

I am on the exercise bike and when I am done I _might_ look at the debug part of this issue but I might also wait until discussion is more complete and just take it easy today too. 

I think it will probably be a day where I don't get anything done or at least not much done. It's one of those days. It doesn't help that I woke up early and feeling very physically exhausted more than usual but it is not strictly that. 

I also don't want to work on something that might potentially change and I have been feeling kind of overwhelmed lately. It's been kind of crazy here. 

Well we shall see. Depending on when you reply to the above comments I might be able to reply today. Otherwise tomorrow morning.

**Comment by lcn2** [comment]
> > What does the JSON parser do if the function:
> 
> > 
> 
> > ```c
> 
> > struct json *json_conv_number(char const *ptr, size_t len)
> 
> > ```
> 
> > 
> 
> > returns NULL?
> 
> > 
> 
> > There MUST/SHOULD/NEEDS be code somewhere in the parser that checks if the "json_conv_foo()" call returns a NULL pointer. Such a code check could also call the `PARSED_JSON_NODE(item)` macro as well and treat that situation also as a "this is invalid JSON" condition.
> 
> 
> 
> The function never returns on NULL so that shouldn't happen?

You are correct: it never returns NULL.

**Comment by xexyl** [comment]
> > > What does the JSON parser do if the function:
> > 
> > > 
> > 
> > > ```c
> > 
> > > struct json *json_conv_number(char const *ptr, size_t len)
> > 
> > > ```
> > 
> > > 
> > 
> > > returns NULL?
> > 
> > > 
> > 
> > > There MUST/SHOULD/NEEDS be code somewhere in the parser that checks if the "json_conv_foo()" call returns a NULL pointer. Such a code check could also call the `PARSED_JSON_NODE(item)` macro as well and treat that situation also as a "this is invalid JSON" condition.
> > 
> > 
> > 
> > The function never returns on NULL so that shouldn't happen?
> 
> You are correct: it never returns NULL.
> 
> 

Yeah. So that part is fine at least. 

I am getting ready for sleep now so I will reply to other comments in the morning if you get to it tonight. 

Hope you have a great evening and sleep well when you do!

**Comment by lcn2** [comment]
With commit 535b11b8156d8a4074ae92f7f41c6a9dfce74610 we compromised on a number of macro issues and fixed a few minor nits.

Hopefully you will see this result as ay least status quo if not a minor improvement.

**Comment by lcn2** [comment]
With commit 5e17b1f2b0709624ab6872a5d869b6ca96835628 we have improved how parsed and converted are used.  We have made a number of improvements on what`vjson_fprint()` prints under `-J 3`.

**Comment by lcn2** [comment]
With commit 2dbf74787806955c3ced598a0a61045cf6e74ee6 the JSON number output has been achieved.

We had to use -J 4 for the extended output as -J 5 started to get too verbose.

See:

```sh
./jparse/jparse -J 3 jparse/test_jparse/test_JSON/good/num-array.json
./jparse/jparse -J 4 jparse/test_jparse/test_JSON/good/num-array.json
```

**Comment by xexyl** [comment]
> With commit [535b11b](https://github.com/ioccc-src/mkiocccentry/commit/535b11b8156d8a4074ae92f7f41c6a9dfce74610) we compromised on a number of macro issues and fixed a few minor nits.
> 
> Hopefully you will see this change at as good at the "status quo" if not a minor improvement.

It seems like I probably will yes: since it works.
> 
> ## UPDATE 0
> For some reason the list if convenience macros in `jparse/json_parse.h` was not added to the change log. They are in `CHANGES.md`. Here they are:
> 
> ```c
> #define PARSED_JSON_NODE(item) ((item)->parsed == true)
> #define CONVERTED_PARSED_JSON_NODE(item) (((item)->parsed == true) && ((item)->converted == true))
> #define VALID_JSON_NODE(item) (((item)->parsed == true) || ((item)->converted == true))
> ```

I noticed that. I know why too. `git` filters out log messages that start out with `#`. I'll put them in the next commit I make if I remember - just have to tab them.

**Comment by xexyl** [comment]
> With commit [2dbf747](https://github.com/ioccc-src/mkiocccentry/commit/2dbf74787806955c3ced598a0a61045cf6e74ee6) the JSON number output has been achieved.
> 
> We had to use -J 4 for the extended output as -J 5 started to get too verbose.
> 
> See:
> 
> ```shell
> ./jparse/jparse -J 3 jparse/test_jparse/test_JSON/good/num-array.json
> ./jparse/jparse -J 4 jparse/test_jparse/test_JSON/good/num-array.json
> ```

That's great! I was going to tackle this but it seems like I don't need to. Thanks!

**Comment by xexyl** [comment]
... I discovered some problems though ... will look at it now ...

**Comment by xexyl** [comment]
> ... I discovered some problems though ... will look at it now ...

.. fixed a few minutes ago .. will commit in a little while ..

**Comment by xexyl** [comment]
> > ... I discovered some problems though ... will look at it now ...
> 
> .. fixed a few minutes ago .. will commit in a little while ..

Done. One commit improved the macros further (check for NULL first) and the other one fixes the problem I noticed: invalid parameter to fprintf call in `JTYPE_UNSET`.

I can look at the other thread in a bit ...

**Comment by xexyl** [comment]
Actually I think I can make some improvements here first: have the flags comma separated. It's very confusing output otherwise. But I'll maybe look at that in a bit .. not sure how much I'm up to yet.

I do have a zoom meeting later on in the morning but I'll be done before that.

**Comment by xexyl** [comment]
Commit https://github.com/ioccc-src/mkiocccentry/pull/795/commits/eb487c72203e353cca92fe3020b014495c5b4881 made some improvements with the parsed/converted flag printing in the `JTYPE_STRING` and `JTYPE_NUMBER` booleans in the debug output. See the log for the reasoning behind it.

Updated version strings.

I'll next look at the other thread but not just yet .. have some other things to do and it might be that I try resting though I seriously doubt I'll be able to just yet so I might look at them soon.

Given you were awake late I hope you're sleeping and well!

**Comment by xexyl** [comment]
.. something broke the conversion routine for numbers ... this causes a problem for me when parsing the lists of the `-n` and `-S` options. Which are all done but I need to test it.

I guess it's related to this issue.

**Comment by xexyl** [comment]
> .. something broke the conversion routine for numbers ... this causes a problem for me when parsing the lists of the `-n` and `-S` options. Which are all done but I need to test it.
> 
> I guess it's related to this issue.

Seems to be related to the changes in the macros. Not sure what you changed. Will have to do later. Will maybe open another issue. Not sure. I'll do a pull request and then will try and rest. Hopefully I can do more later.

**Comment by xexyl** [comment]
> > .. something broke the conversion routine for numbers ... this causes a problem for me when parsing the lists of the `-n` and `-S` options. Which are all done but I need to test it.
> > I guess it's related to this issue.
> 
> Seems to be related to the changes in the macros. Not sure what you changed. Will have to do later. Will maybe open another issue. Not sure. I'll do a pull request and then will try and rest. Hopefully I can do more later.


The problem was indeed related to a change you made. This diff makes it work:

```diff
-#define PARSED_JSON_NODE(item) ((item) != NULL && ((item)->parsed == true))
+#define PARSED_JSON_NODE(item) ((item)->parsed && !(item)->converted)
```

Either I need another macro or some other thing has to change but not having it this way breaks the check for just parsed being true and not converted. This is because we have to check that converted failed but parsed did not. Otherwise how would we know when conversion failed but parsing is okay?

But since you seem set on the macro not being the way I had it I'll let you reply first ... still the use is to make sure that ONLY parsed is true.

.. resting now .. back later today maybe else tomorrow.

**Comment by xexyl** [comment]
> ```diff
> #define PARSED_JSON_NODE(item) ((item)->parsed && !(item)->converted)
> ```

Fixed in commit 54327025783a801f8f931dadf9d3518339aae396.

**Comment by xexyl** [comment]
> > > .. something broke the conversion routine for numbers ... this causes a problem for me when parsing the lists of the `-n` and `-S` options. Which are all done but I need to test it.
> > > I guess it's related to this issue.
> > 
> > 
> > Seems to be related to the changes in the macros. Not sure what you changed. Will have to do later. Will maybe open another issue. Not sure. I'll do a pull request and then will try and rest. Hopefully I can do more later.
> 
> The problem was indeed related to a change you made. This diff makes it work:
> 
> ```diff
> -#define PARSED_JSON_NODE(item) ((item) != NULL && ((item)->parsed == true))
> +#define PARSED_JSON_NODE(item) ((item)->parsed && !(item)->converted)
> ```
> 
> Either I need another macro or some other thing has to change but not having it this way breaks the check for just parsed being true and not converted. This is because we have to check that converted failed but parsed did not. Otherwise how would we know when conversion failed but parsing is okay?
> 
> But since you seem set on the macro not being the way I had it I'll let you reply first ... still the use is to make sure that ONLY parsed is true.
> 
> .. resting now .. back later today maybe else tomorrow.

Fixed in commit https://github.com/ioccc-src/mkiocccentry/commit/54327025783a801f8f931dadf9d3518339aae396.
